tutorials/notebook/index_js.md

```diff
@@ -0,0 +1,175 @@
+---
+layout: default
+title: index.js
+parent: Notebook
+nav_order: 2
+---
+
+```javascript
+const tupelo = require('tupelo-client');
+const fs = require('fs');
+const yargs = require('yargs');
+
+const LOCAL_ID_PATH = './.timestamper-identifiers';
+const CHAIN_TREE_NOTE_PATH='notebook/notes';
+
+function connect(creds) {
+  return tupelo.connect('localhost:50051', creds);
+}
+
+function identifierObj(key, chain) {
+  return {
+    keyAddr: key,
+    chainId: chain
+  };
+}
+
+function writeIdentifierFile(configObj) {
+  let data = JSON.stringify(configObj);
+  fs.writeFileSync(LOCAL_ID_PATH, data);
+}
+
+function readIdentifierFile() {
+  let raw = fs.readFileSync(LOCAL_ID_PATH);
+  return JSON.parse(raw);
+}
+
+function idFileExists() {
+  return fs.existsSync(LOCAL_ID_PATH);
+}
+
+function createNotebook(creds) {
+  let client = connect(creds);
+  var keyAddr, chainId;
+
+  client.register()
+    .then(function(registerResult){
+      return client.generateKey();
+    }, function(err) {
+      console.log("Error registering wallet.");
+      console.log(err);
+    }).then(function(generateKeyResult) {
+      keyAddr = generateKeyResult.keyAddr;
+      return client.createChainTree(keyAddr);
+    }, function(err) {
+      console.log("Error generating key.");
+      console.log(err);
+    }).then(function(createChainResponse) {
+      chainId = createChainResponse.chainId;
+      console.log("Saving registration.");
+      let obj = identifierObj(keyAddr, chainId);
+      return writeIdentifierFile(obj);
+    }, function(err) {
+      console.log("Error creating chain tree.");
+      console.log(err);
+    });
+}
+
+function addTimestamp(note) {
+  let ts = new Date().getTime().toString();
+  return ts + '::' + note;
+}
+
+function addNote(creds, note) {
+  if (!idFileExists()) {
+    console.log("Error: you must register before you can record notes.");
+
+    return;
+  }
+
+  let client = connect(creds),
+      identifiers = readIdentifierFile();
+
+  client.resolve(identifiers.chainId, CHAIN_TREE_NOTE_PATH)
+    .then(function(resp) {
+          let notes = resp.data,
+        noteWithTs = addTimestamp(note);
+
+      if (notes instanceof Array) {
+        notes.push(noteWithTs);
+      } else {
+        notes = [noteWithTs];
+      }
+
+      return client.setData(identifiers.chainId,
+                            identifiers.keyAddr,
+                            CHAIN_TREE_NOTE_PATH,
+                            notes);
+  }, function(err) {
+      console.log('Error reading notes: ' + err);
+  });
+}
+
+function showNotes(creds) {
+  if (!idFileExists()) {
+    console.log("Error: you must register before you can print notes.");
+
+    return;
+  }
+
+  let identifiers = readIdentifierFile();
+  let client = connect(creds);
+  let path = CHAIN_TREE_NOTE_PATH;
+
+  client.resolve(identifiers.chainId, CHAIN_TREE_NOTE_PATH)
```

Comment:
`s/CHAIN_TREE_NOTE_PATH/path/`?

Reply:
@cap10morgan the incremental testing was a good suggestion. I made some changes to the tupelo binary and changed the tutorial so that we can create notebooks multiple times. Basically, I return a descriptive error with the correct error code (`AlreadyExists` instead of `Unknown`) from the Tupelo side, then I look for and ignore that error from the tutorial side for subsequent calls to `client.register()`. That way, people can keep running the code over and over without the tutorial blowing up. I then changed the tutorial so that the reader periodically tests the code they have so far at the node repl. 

I'm going to go ahead and merge since I think this is at least a good version 1 that would be helpful to other devs, but if you'd still like to see changes, then feel free to either ping me here or open an issue and I'll work on those changes for version 2.